src/cdomains/intDomain.ml

```diff
@@ -923,6 +923,461 @@ struct
+
+  let widen ik xs ys = 
+    let (min_ik,max_ik) = range ik in interval_sets_to_partitions ik None xs ys |> merge_list ik |> 
+    (function
+    | [] -> []
+    | (None,(lb,rb))::ts -> (None, (min_ik,rb))::ts
+    | (Some (la,ra), (lb,rb))::ts  when Ints_t.compare lb la < 0 -> (Some (la,ra),(min_ik,rb))::ts
+    | x  -> x)
+    |> List.rev
+    |> (function
+    | [] -> []
+    | (None,(lb,rb))::ts -> (None, (lb,max_ik))::ts
+    | (Some (la,ra), (lb,rb))::ts  when Ints_t.compare ra rb < 0 -> (Some (la,ra),(lb,max_ik))::ts
+    | x  -> x)
+    |> List.rev
+    |> List.map snd  
+
+  let starting ik n = norm ik @@ Some (n, snd (range ik))
+
+  let ending ik n = norm ik @@ Some (fst (range ik), n)
+
+  let minimal = function [] -> None | (x, _)::_ -> Some x
+
+  let maximal xs = xs |> List.rev |> (function [] -> None | (_, y)::_ -> Some y)
+
+  let of_interval ik (x, y) = norm ik @@ Some (x, y)
+
+  let of_int ik (x: int_t) = of_interval ik (x, x)
+
+  let of_bool _ik = function true -> one | false -> zero
+  
+  let invariant_ikind_interval e ik x = match x with
+  | (x1, x2) when Ints_t.compare x1 x2 = 0 ->
+    let x1 = Ints_t.to_bigint x1 in
+    Invariant.of_exp Cil.(BinOp (Eq, e, kintegerCilint ik x1, intType))
+  | (x1, x2) ->
+    let open Invariant in
+    let (min_ik, max_ik) = range ik in
+    let (x1', x2') = BatTuple.Tuple2.mapn (Ints_t.to_bigint) (x1, x2) in
+    let i1 = if Ints_t.compare min_ik x1 <> 0 then of_exp Cil.(BinOp (Le, kintegerCilint ik x1', e, intType)) else none in
+    let i2 = if Ints_t.compare x2 max_ik <> 0 then of_exp Cil.(BinOp (Le, e, kintegerCilint ik x2', intType)) else none in
+    i1 && i2
+
+  let invariant_ikind e ik xs = List.map (invariant_ikind_interval e ik) xs |> let open Invariant in List.fold_left (||) (bot ())
+  
+  let modulo n k =
+    let result = Ints_t.rem n k in
+    if Ints_t.compare result Ints_t.zero >= 0 then result
+    else Ints_t.add result k
+
+  let refine_with_congruence_interval ik (cong : (int_t * int_t ) option) (intv : (int_t * int_t ) option): t =
+    match intv, cong with
+    | Some (x, y), Some (c, m) ->
+      if Ints_t.equal m Ints_t.zero && (Ints_t.compare c x < 0 || Ints_t.compare c y > 0) then []
+      else if Ints_t.equal m Ints_t.zero then
+        [(c, c)]
+      else
+        let (min_ik, max_ik) = range ik in
+        let rcx =
+          if Ints_t.equal x min_ik then x else
+            Ints_t.add x (modulo (Ints_t.sub c x) (Ints_t.abs m)) in
+        let lcy =
+          if Ints_t.equal y max_ik then y else
+            Ints_t.sub y (modulo (Ints_t.sub y c) (Ints_t.abs m)) in
+        if Ints_t.compare rcx lcy > 0 then []
+        else if Ints_t.equal rcx lcy then norm ik @@ Some (rcx, rcx)
+        else norm ik @@ Some (rcx, lcy)
+    | _ -> []
+
+  let refine_with_congruence ik (intvs: t) (cong: (int_t * int_t ) option): t =
+    List.map (fun x -> Some x) intvs |> List.map (refine_with_congruence_interval ik cong) |> List.flatten
```

Comment:
There is no reason to pollute the module with `refine_with_congruence_interval` here, it should be defined locally inside `refine_with_congruence`.